File: libraries/types/audio.lua
local engine = require('libraries.engine')

monoe = monoe or {}

---@class monoe.audio
---@field uid integer
monoe.audio = {}
monoe.audio.__index = monoe.audio

local base = 'monoe.exe.Core.Bridge.Types.Audio'

---creates a new audio player
---@param source string?
function monoe.audio.new(source)
  local uid = engine.import(base)

  if uid == -1 then
    error('got invalid UID when create an instance of ' .. base)
  end

  if type(source) == "string" then
    engine.call(uid, 'Load', source, source:sub(-3))
  end

  return setmetatable({ uid = uid }, monoe.audio)
end

---loads a sound (.wav or .mp3)
---@param source string
function monoe.audio:load(source)
  engine.call(self.uid, 'Load', source, source:sub(-3))
end

---plays the audio
---@param at number?
---@param loop boolean?
function monoe.audio:play(at, loop)
  engine.call(self.uid, 'Play', at, loop or false)
end

function monoe.audio:stop()
  engine.call(self.uid, 'Stop')
end

---returns the size of the stream
---@return number
function monoe.audio:length()
  return engine.call(self.uid, 'Length')
end

---returns the name of a new event 
---@return string
function monoe.audio:finished()
  return engine.call(self.uid, 'FinishedEvent')
end

function monoe.audio:free()
  engine.call(self.uid, 'Free')
end

_G.monoe = monoe
_G.monoe.audio = monoe.audio
return monoe.audio

File: Core/Bridge/Types/Audio.cs
using System.IO;
using Godot;

namespace monoe.exe.Core.Bridge.Types;

public class Audio : Exposable
{
  private readonly AudioStreamPlayer audio = new();
  private bool isEventAdded = false;

  public void Load(string path, string fmt)
  {
    var file = Godot.FileAccess.Open(path, Godot.FileAccess.ModeFlags.Read) ?? throw new FileNotFoundException($"cannot open file {path}");

    var buff = file.GetBuffer((long)file.GetLength());

    audio.Stream = fmt switch
    {
      "wav" => new AudioStreamWav() { Data = buff },
      "mp3" => new AudioStreamMP3() { Data = buff },
      _ => throw new InvalidDataException($"unsupported format {fmt}"),
    };
  }

  public void Play(double at = 0.0, bool loop = false)
  {
    audio.Play((float)at);
    if (audio.Stream is AudioStreamWav wav)
      wav.LoopMode = AudioStreamWav.LoopModeEnum.Forward;
    else if (audio.Stream is AudioStreamMP3 mP3)
      mP3.Loop = loop;
  }

  public void Stop() => audio.Stop();
  public double Length() => audio.Stream.GetLength();
  
  public string FinishedEvent()
  {
    string @event = $"_on_audio#{UID}_finished";

    if (isEventAdded) audio.Finished += () =>
    {
      Main.Emit(@event);
    };

    isEventAdded = true;
    return @event;
  }

  public override Node NRef() => audio;

  protected override void _Free()
  {
    audio.QueueFree();
  }
}